### DISCUSSION TITLE
Have breadcrumbs link to the first document in a category if the category itself is not a link

### DISCUSSION THREAD

**Discussion by patrys** [facebook/docusaurus#8446] [Ideas]
As the title says, it would be nice to have the option of breadcrumbs always linking somewhere.

We have big API reference pages, and it does not make sense to try to generate index pages for them. But it also means that breadcrumbs appear broken as nothing is clickable.

A similar problem appears if you try to autogenerate sidebars from a multi-level directory structure, it correctly populates the sidebar but makes breadcrumbs unusable as none of the categories are pages themselves.

**Comment by patrys** [comment]
Here's an example of docs generated using [GraphQL-Markdown](https://graphql-markdown.github.io):

https://docs.saleor.io/docs/3.x/api-reference/enums/address-type-enum

**Reply by slorber** [reply]
So what I understand is that you don't want a page for Enums. 
But you want Enums to be clickable and link to AccountsErrorCode that's right?

![CleanShot 2022-12-15 at 17 30 02@2x](https://user-images.githubusercontent.com/749374/207915802-24f93d6b-d9f1-4233-8ce3-8753df99844b.png)

Technically you can use a doc link for that on the category sidebar item: `link: {type: 'doc', id: 'AccountsErrorCode'}`

I wouldn't be against an option to make it easier, such as `link: {type: "first-doc}`

Note that when JS is disabled, we already add a link to the Enums category (as it allows to "expand" the category by simply navigating, otherwise it's not possible without JS) and technically we already have the logic to find the first doc to link to.

Applying all this globally as a default to your site is a whole new extra API to design. It also makes sense to allow all categories to have a generated index by default, and opt-in to have another kind of link or simply disable linking.

**Reply by slorber** [reply]
> A similar problem appears if you try to autogenerate sidebars from a multi-level directory structure, it correctly populates the sidebar but makes breadcrumbs unusable as none of the categories are pages themselves.

Note: intermediate breadcrumb items are just sidebar categories. If you make sidebar categories clickable, the breadcrumbs will be clickable as well. All the solutions above focus on making the categories clickable, as both are linked. You can as well drop a `index.md` in a subfolder and the category would become clickable.

**Reply by patrys** [reply]
What would you put in an index file for the example I linked? It does not make sense to generate cards for all of the entries.

As the sidebar uses an auto-generated tree, I also don’t see a way to add a `link` unless it’s supported by `_category_.json`.

**Reply by slorber** [reply]
> What would you put in an index file for the example I linked?

You put whatever you want in this file, it's up to you. You can write an intro text for Enums without rendering cards.

>  It does not make sense to generate cards for all of the entries.

It doesn't make sense either to have the category to link to its first doc in all cases. In particular for deeply nested categories this might be confusing to have the top-level category to link to a deeply nested doc (ie the first doc being in a sidebar tree). At least it's not a very intuitive default behavior that's why I prefer it to be opt-in.




---

> As the sidebar uses an auto-generated tree, I also don’t see a way to add a link unless it’s supported by _category_.json.

Indeed, it is supported:

https://docusaurus.io/docs/next/sidebar/autogenerated#category-item-metadata

![CleanShot 2022-12-16 at 13 30 19@2x](https://user-images.githubusercontent.com/749374/208098664-d9241486-2734-4007-8cac-4a098e97b2e5.png)

**Reply by patrys** [reply]
Thanks, I was not aware of that! In this case the only missing bit is being able to say “first child” without having to name a specific document.

**Reply by slorber** [reply]
Technically you can customize the sidebar generator so that it automatically links all categories it generates to their first docs child :)

https://docusaurus.io/docs/next/sidebar/autogenerated#customize-the-sidebar-items-generator


```js
function linkCategoriesToFirstChild(items) {
  // TODO do whatever you want here
}

module.exports = {
  plugins: [
    [
      '@docusaurus/plugin-content-docs',
      {
        async sidebarItemsGenerator({defaultSidebarItemsGenerator, ...args}) {
          const sidebarItems = await defaultSidebarItemsGenerator(args);
          return linkCategoriesToFirstChild(sidebarItems);
        },
      },
    ],
  ],
};
```
